### PR TITLE
ENH: Multiple color bars via displayable manager support

### PR DESCRIPTION
**Pull request by MichaelColonel**
This is a WIP of the issue #4891.

Color bar will be controlled via vtkMRMLColorBarDisplayNode. Display node can observe either vtkMRMLColorTableNode directly or vtkMRMLDisplayableNode and then get a color table node for color bar visualization.

If there are no obvious errors or mistakes, the next step will be modification of the "Colors" module GUI in order to change properties of the selected color bar actor.

### CODE REVIEW

**Comment by MichaelColonel** [comment]
Most of the work is done and ready for a review. I will rework tests later.

How to test:
1. Select a node in a "Display Color Bar for Node" collapsible button.
2. Click an eye icon to show a color bars.

To modify some properties of a vtkScalarBar select "Scalar bar from view " node in a "Color Scalar Bar"collapsible button.

image

**Comment by MichaelColonel** [comment]
Has been added a test to check that color bar is visible on slice and 3D views.

File: Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx
Comment (lassoan) [inline Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx:243]: Use scene->AddNewNodeByClass instead of vtkNew+AddNode, because that is shorter and it also allows user to override default node properties (using the scene's default node mechanism).

File: Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx
Comment (lassoan) [inline Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx:248]: The color bar is just another display node (similarly how the scalar volume node has additional display node(s) for volume rendering), so we don't need a special reference role. Instead, we should call:

    dispNode->AddAndObserveDisplayNodeID(colorBarNode->GetID());

File: Modules/Loadable/Colors/MRML/vtkMRMLColorBarDisplayNode.cxx
Comment (lassoan) [inline Modules/Loadable/Colors/MRML/vtkMRMLColorBarDisplayNode.cxx:34]: no need for `displayableRef` becaues display nodes already maintain a pointer to their associated displayable node by other means

File: Modules/Loadable/Colors/MRML/vtkMRMLColorBarDisplayNode.cxx
Comment (lassoan) [inline Modules/Loadable/Colors/MRML/vtkMRMLColorBarDisplayNode.cxx:35]: Shouldn't we use the displayable node's color node?

File: Modules/Loadable/Colors/MRMLDM/vtkMRMLColorBarDisplayableManager.cxx
Comment (lassoan) [inline Modules/Loadable/Colors/MRMLDM/vtkMRMLColorBarDisplayableManager.cxx:96]: We should observe all vtkMRMLColorBarDisplayNodes here and add a vtkScalarBarWidget/vtkSlicerScalarBarActor for each that are displayable in the view. See vtkMRMLVolumeRenderingDisplayableManager and markups displayable manager for example.

**Review by lassoan — CHANGES_REQUESTED** [review]
Thank you, this is very nice progress. I've added a few comments inline, but the most important change that I think we would need to do is to support any number of scalar bars - one for each scalar bar display node - in a view. The scalar bar would be enabled/disabled in each displayable node's GUI (in Volumes, Models, Markups, etc. modules).

A Qt widget could be added for editing the scalar bar options (font, style, etc.), which would be all stored in the scalar bar display node.

File: Modules/Loadable/Colors/MRMLDM/vtkMRMLColorBarDisplayableManager.cxx
Comment (MichaelColonel) [reply Modules/Loadable/Colors/MRMLDM/vtkMRMLColorBarDisplayableManager.cxx:96]: Got it! I will add support for multiple scalar bars, one for each displayable node.

File: Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx
Comment (MichaelColonel) [reply Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx]: Fixed. Everything else gives pre-processor  error.

File: Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx
Comment (MichaelColonel) [reply Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx]: Fixed.

File: Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx
Comment (MichaelColonel) [reply Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx]: Fixed.

File: Modules/Loadable/Colors/MRML/vtkMRMLColorBarDisplayNode.cxx
Comment (MichaelColonel) [reply Modules/Loadable/Colors/MRML/vtkMRMLColorBarDisplayNode.cxx]: Fixed.

**Comment by MichaelColonel** [comment]
Added initial support of multiple color bars on the single view. Test is updated as well.

PR is ready for a review once again.

image

**Comment by MichaelColonel** [comment]
> edit display properties in Volumes and Models module (in addition to Colors module). For that the best would be to have a reusable widget that can be easily added to all these modules.

Would you recommend qMRMLDisplayNodeWidget for that or something else?

**Comment by lassoan** [comment]
Thank you for working on this. Let us know if it is ready for review and testing.

**Comment by MichaelColonel** [comment]
I think the PR is ready for review and testing.

Color bar control widget was added into Volumes, Colors and Models modules. Color bar test was modified once again.

image

There is problem with color bar for model nodes. If i add color bar to the model with already visible model display node the program crashes. If color bar is added to model with disabled model display node the visibility icon in hierarchy widget looks like
image
and when i click on the icon to make is visible the crashes too.

Everything else should work.

**Comment by MichaelColonel** [comment]
Color bar display node is ready for review and testing. Color bar display node test modified as well.

1. "Color Bar" group box with qMRMLColorBarDisplayWidget was added to "Models" module widget. 
2. "Color Scalar Bar" collapsible button with qMRMLColorBarDisplayWidget was added to "Volumes" module widget.
3. "Color Scalar Bar" collapsible button with qMRMLColorBarDisplayWidget was added to "Colors" module widget.

Context menu items for color bar will be added shortly, if everything else works without problem.

**Comment by MichaelColonel** [comment]
"Color Scalar Bar" settings within "Volumes" module
image

"Color Bar" settings within "Models" module
image

**Comment by lassoan** [comment]
This looks awesome, thank you very much! Sorry for the review taking so long, I hope I can give feedback (or merge as is if everything looks good as is) early next week.

**Comment by lassoan** [comment]
I've tested this and it works very well. Thanks a lot for implementing this. I have a couple of comments that would need to be addressed before starting a final review and integration:

High priority (fix required before integration):
- [ ] Displayed range is not in sync with the colorbar range. This would be one of the huge benefits of this new color bar design: user would automatically see the actual scalar value->color mapping. For this the color bar display node must be able to keep a reference to a primary display node (e.g., vtkMRMLModelDisplayNode, vtkMRMLVolumeDisplayNode) and if that reference is set then get the effective display range from there. The primary display node knows if the display range comes from the color node, automatically set based on the data range, or it the user set that manually. If I change window/level of the volume then the color bar should update. Similarly, as the display range is adjusted in Models module / Display / Scalars / Displayed Range, then the numbers on the color bar should be automatically updated.

GUI tweaks (should be fixed before the integration, because should not take a lot of work and would improve usability):
  - [ ] If I keep clicking on the "Vertical" and "Horizontal" radiobutton then the width and height components are keep flipping
  - [ ] If editing the Text or Format field of the color bar then the cursor jumps to the end (you can add a check to only update the widget text from MRML if they are different)
  - [ ] Add tooltip for format string (mention that it is a [printf](https://www.cplusplus.com/reference/cstdio/printf/) format string and add a few examples and short explanation of the most important options)
  - [ ] Hide `Labels properties` / `Size` slider (it is now always disabled, which is a bit confusing)
  - [ ] Position with in view and size -> show as percentage (instead of a number between 0.0-1.0)
  - [ ] Make the scalar bar position lower-left by default so that the same position works well for both horizontal and vertical orientation
  - [ ] In Volumes module, move the Color Scalar Bar section above the spacer (it is hard to find it now at the very bottom)

Medium priority (would be nice to address in this pull request, but not required):
- [ ] If according to the primary display node a model (or volume, etc.) node is not shown in a particular view then don't show the scalar bar there either. Only show that if the displayable node is visible there _and_ visibility of the color bar display node is enabled there, too.
- [ ] Title properties / Text: it would be nice to allow placeholders [node] and [scalar] that would be replaced by the node's name and the active scalar's name (default text could be "[node] [scalar]")

Low priority (should be done after this pull request is integrated):
- [ ] Interactive positioning/sizing could be nice
- [ ] Remove Color Scalar bar from DataProbe module - it is not needed anymore

Thanks again all your hard work on this. I'm very impressed that you were able to implement such a complex feature, and the Slicer community will really appreciate these results, too.

**Comment by MichaelColonel** [comment]
> If I change window/level of the volume then the color bar should update. Similarly, as the display range is adjusted in Models module / Display / Scalars / Displayed Range, then the numbers on the color bar should be automatically updated.

So color bar must display scalar range from primary display node? For example for CT with WL = 300, WW = 200 the color bar labels must be from 200 up to 400, or labels can be colors values but in the right range?

> In Volumes module, move the Color Scalar Bar section above the spacer (it is hard to find it now at the very bottom)

The color bar collapsible button with color bar widget is already above the spacer, the problem is a volume display widget that above the color bar widget (i has a spacer which push color bar widget to the bottom).

**Comment by MichaelColonel** [comment]
> Title properties / Text: it would be nice to allow placeholders [node] and [scalar] that would be replaced by the node's name and the active scalar's name (default text could be "[node] [scalar]")

Should the title be changed every time scalar node name is changed or title is set only once in the beginning?

**Comment by lassoan** [comment]
> Should the title be changed every time scalar node name is changed or title is set only once in the beginning?

It would need to be kept up-to-date (update whenever the node name or active scalar name is changed).

**Comment by MichaelColonel** [comment]
High and medium priority changes are fulfilled in general, and the PR is ready for another review.

**Comment by lassoan** [comment]
Thanks a lot! I'm still struggling with integration/design update of the interactive slice intersection pull request, so I think I'll only get to review and integrate this pull request on Monday.

**Comment by lassoan** [comment]
I'm refactoring the code and it should be ready early next week. Don't make any more changes until I finish with it. Thanks for your patience.

**Review by lassoan — APPROVED** [review]
I've made a number of improvements, this should be ready to be merged. I'll just wait for the automatic tests to complete and see if anyone gives any feedback.

**Comment by lassoan** [comment]
As there are many modified files and many another pull request to process in the following days, I would like to get this merged soon. 

All checks have passed, there are no regressions in automated tests, tests added, documentation updated, manually tested and everything looks good.

If I don't hear from anybody then I'll merge it this afternoon.

**Comment by MichaelColonel** [comment]
Once it will be merged i will try to fix low priority issues.

**Comment by lassoan** [comment]
Merged. I've added a ticket to track further improvements/low-priority issues: https://github.com/Slicer/Slicer/issues/6111

**Comment by lassoan** [comment]
@MichaelColonel this all works very nicely. I think it is time to announce the new feature on the Slicer Forum. Could you please add a new topic in "Support" category, add the tag "feature", describing the new feature from the user standpoint and attach a few images, maybe a short video, too. If you want to add any notes for developers then add it after describing the feature from the user perspective. Invite people to give comments and suggestions at the very end of the post.

**Pull request merged**
